# Exercise 1

In this exercise, we work with the California Housing dataset, which is included in sklearn. It is a collection of average house prices per block, while each block is also characterized by 8 features:

|Feature name | description|
|-------------|------------|
|MedInc | median income in block|
|HouseAge | median house age in block|
|AveRooms | average number of rooms|
|AveBedrms | average number of bedrooms|
|Population | block population|
|AveOccup | average house occupancy|
|Latitude | house block latitude|
|Longitude | house block longitude|

We can download the data with:

In [3]:
from sklearn.datasets import fetch_california_housing

(X, y) = fetch_california_housing(data_home=None, download_if_missing=True, return_X_y=True)

X and y are the features and the label, respectively. We can look at the format of the data:

In [7]:
print(type(X))
print(type(y))

print(X.shape)
print(y.shape)

print(X.dtype)
print(y.dtype)


<class 'numpy.ndarray'>
<class 'numpy.ndarray'>
(20640, 8)
(20640,)
float64
float64


As you can see, X and y are numpy arrays. There are a total of 20640 samples in the dataset. X has 8 features per sample, and y are the corresponding labels.

## Plotting the data

When working with a dataset, it is always helpful to visualize the data that we are working with, in order to be able to check if our results are meaningful. We will use matplotlib to plot the data.

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
features = ["MedInc", "HouseAge", "AveRooms", "AveBedrms", "Population", "AveOccup", "Latitude", "Longitude"]

for i, feature_name in enumerate(features):
    plt.figure()
    plt.title(f"price vs {feature_name}")
    plt.scatter(X[:, i], y, marker="x")
    plt.show()

We can see that the data is non-linear, which means that a linear regression will probably not yield the best results. But we can do it anyway and see how good it performs.

Also, this dataset has more dimensions than the example, which was 2D. But linear regression works still the same. Instead of one input variable we now have eight input variables, to calculate the output (the price). The linear regression will find a factor $\alpha_{var}$ for each input variable $var$ plus a bias $b$. This means, we assume that the price can be calculated in the following form:

\begin{equation}
  price = \alpha_{MedInc} MedInc + \alpha_{HouseAge} HouseAge + \ldots + \alpha_{Longitude} Longitude + b
\end{equation}

Your tasks in this notebook are:
    
- Add and adapt the numpy linear regression code from the other regression notebook, so that it performs linear regression on the California Housing dataset. If performed correctly, you should be able to get an MSE of ca. 0.5243. You can use sklearn to get a quick solution, but also implement the manual solution (without sklearn).

## Solution